File: auth-web/lib/qualityPolicy.ts
export const QUALITY_POLICY_KEY = 'hermes_generation'

export const DEFAULT_QUALITY_POLICY = {
    schema_version: 1,
    topic: { enabled: true, min_title_chars: 12 },
    plan: { enabled: true, min_scenes: 1, require_media_status_ready: true },
    script: {
        enabled: true, min_quality_score: 78, min_hangul_chars: 1000,
        max_latin_ratio: 0.05, max_repeated_paragraph_opener: 2,
        prohibit_fallback: true, prohibit_off_category: true,
    },
    media: {
        enabled: true, min_image_prompt_chars: 120, min_video_prompt_chars: 260,
        max_video_prompt_scenes: 12, required_camera_movements: 1,
        require_video_guardrails: true, prohibit_duplicate_prompts: true,
        require_image_grids: true,
    },
    publish: {
        enabled: true, min_description_chars: 120,
        require_language_match: true, prohibit_internal_terms: true,
    },
    delivery: {
        enabled: true, require_all_prior_stages: true,
        require_quality_report_pass: true, block_scene_count_mismatch: true,
    },
} as const

type JsonObject = Record<string, unknown>
const object = (value: unknown): JsonObject => value && typeof value === 'object' && !Array.isArray(value) ? value as JsonObject : {}
const bool = (value: unknown, fallback: boolean) => typeof value === 'boolean' ? value : fallback
const number = (value: unknown, fallback: number, min: number, max: number) => {
    const parsed = typeof value === 'number' ? value : Number(value)
    return Number.isFinite(parsed) ? Math.max(min, Math.min(max, parsed)) : fallback
}

export function normalizeQualityPolicy(value: unknown) {
    const root = object(value)
    const topic = object(root.topic)
    const plan = object(root.plan)
    const script = object(root.script)
    const media = object(root.media)
    const publish = object(root.publish)
    const delivery = object(root.delivery)
    return {
        schema_version: 1,
        topic: { enabled: bool(topic.enabled, true), min_title_chars: number(topic.min_title_chars, 12, 1, 300) },
        plan: {
            enabled: bool(plan.enabled, true), min_scenes: number(plan.min_scenes, 1, 1, 500),
            require_media_status_ready: bool(plan.require_media_status_ready, true),
        },
        script: {
            enabled: bool(script.enabled, true), min_quality_score: number(script.min_quality_score, 78, 0, 100),
            min_hangul_chars: number(script.min_hangul_chars, 1000, 0, 200000),
            max_latin_ratio: number(script.max_latin_ratio, 0.05, 0, 1),
            max_repeated_paragraph_opener: number(script.max_repeated_paragraph_opener, 2, 1, 100),
            prohibit_fallback: true,
            prohibit_off_category: bool(script.prohibit_off_category, true),
        },
        media: {
            enabled: bool(media.enabled, true), min_image_prompt_chars: number(media.min_image_prompt_chars, 120, 1, 10000),
            min_video_prompt_chars: number(media.min_video_prompt_chars, 260, 1, 10000),
            max_video_prompt_scenes: number(media.max_video_prompt_scenes, 12, 0, 500),
            required_camera_movements: number(media.required_camera_movements, 1, 0, 10),
            require_video_guardrails: bool(media.require_video_guardrails, true),
            prohibit_duplicate_prompts: bool(media.prohibit_duplicate_prompts, true),
            require_image_grids: bool(media.require_image_grids, true),
        },
        publish: {
            enabled: bool(publish.enabled, true), min_description_chars: number(publish.min_description_chars, 120, 0, 10000),
            require_language_match: bool(publish.require_language_match, true),
            prohibit_internal_terms: bool(publish.prohibit_internal_terms, true),
        },
        delivery: {
            enabled: bool(delivery.enabled, true), require_all_prior_stages: true,
            require_quality_report_pass: true,
            block_scene_count_mismatch: bool(delivery.block_scene_count_mismatch, true),
        },
    }
}

File: services/quality_policy.py
"""Validated runtime quality policy shared by Hermes quality gates."""

from __future__ import annotations

from copy import deepcopy
from typing import Any, Mapping

POLICY_KEY = "hermes_generation"
DEFAULT_QUALITY_POLICY: dict[str, Any] = {
    "schema_version": 1,
    "topic": {"enabled": True, "min_title_chars": 12},
    "plan": {"enabled": True, "min_scenes": 1, "require_media_status_ready": True},
    "script": {
        "enabled": True, "min_quality_score": 78, "min_hangul_chars": 1000,
        "max_latin_ratio": 0.05, "max_repeated_paragraph_opener": 2,
        "prohibit_fallback": True, "prohibit_off_category": True,
    },
    "media": {
        "enabled": True, "min_image_prompt_chars": 120, "min_video_prompt_chars": 260,
        "max_video_prompt_scenes": 12, "required_camera_movements": 1,
        "require_video_guardrails": True, "prohibit_duplicate_prompts": True,
        "prohibit_duplicate_scene_summaries": True,
        "prohibit_duplicate_retention_hooks": True, "require_image_grids": True,
    },
    "publish": {
        "enabled": True, "min_description_chars": 120,
        "require_language_match": True, "prohibit_internal_terms": True,
    },
    "delivery": {
        "enabled": True, "require_all_prior_stages": True,
        "require_quality_report_pass": True, "block_scene_count_mismatch": True,
    },
}
_active_policy = deepcopy(DEFAULT_QUALITY_POLICY)
_active_version = 0


def _merge_known(defaults: Mapping[str, Any], candidate: Mapping[str, Any]) -> dict[str, Any]:
    merged: dict[str, Any] = {}
    for key, default in defaults.items():
        value = candidate.get(key, default)
        if isinstance(default, Mapping):
            merged[key] = _merge_known(default, value if isinstance(value, Mapping) else {})
        elif isinstance(default, bool):
            merged[key] = value if isinstance(value, bool) else default
        elif isinstance(default, int):
            merged[key] = value if isinstance(value, int) and not isinstance(value, bool) else default
        elif isinstance(default, float):
            merged[key] = float(value) if isinstance(value, (int, float)) and not isinstance(value, bool) else default
        else:
            merged[key] = value if isinstance(value, type(default)) else default
    return merged


def normalize_quality_policy(candidate: Mapping[str, Any] | None) -> dict[str, Any]:
    policy = _merge_known(DEFAULT_QUALITY_POLICY, candidate or {})
    policy["script"]["min_quality_score"] = max(0, min(100, policy["script"]["min_quality_score"]))
    policy["script"]["min_hangul_chars"] = max(0, policy["script"]["min_hangul_chars"])
    policy["script"]["max_latin_ratio"] = max(0.0, min(1.0, policy["script"]["max_latin_ratio"]))
    policy["script"]["max_repeated_paragraph_opener"] = max(1, policy["script"]["max_repeated_paragraph_opener"])
    policy["plan"]["min_scenes"] = max(1, policy["plan"]["min_scenes"])
    policy["media"]["min_image_prompt_chars"] = max(1, policy["media"]["min_image_prompt_chars"])
    policy["media"]["min_video_prompt_chars"] = max(1, policy["media"]["min_video_prompt_chars"])
    policy["media"]["max_video_prompt_scenes"] = max(0, policy["media"]["max_video_prompt_scenes"])
    policy["media"]["required_camera_movements"] = max(0, policy["media"]["required_camera_movements"])
    policy["publish"]["min_description_chars"] = max(0, policy["publish"]["min_description_chars"])
    policy["script"]["prohibit_fallback"] = True
    policy["media"]["prohibit_duplicate_prompts"] = True
    policy["media"]["prohibit_duplicate_scene_summaries"] = True
    policy["media"]["prohibit_duplicate_retention_hooks"] = True
    policy["delivery"]["require_all_prior_stages"] = True
    policy["delivery"]["require_quality_report_pass"] = True
    return policy


def set_active_quality_policy(candidate: Mapping[str, Any] | None, version: int = 0) -> dict[str, Any]:
    global _active_policy, _active_version
    _active_policy = normalize_quality_policy(candidate)
    _active_version = max(0, int(version or 0))
    return quality_policy_snapshot()


def quality_policy_snapshot() -> dict[str, Any]:
    return {"key": POLICY_KEY, "version": _active_version, "policy": deepcopy(_active_policy)}


def active_quality_policy() -> dict[str, Any]:
    return deepcopy(_active_policy)
